src/oscar/apps/communication/abstract_models.py

```diff
@@ -0,0 +1,217 @@
+from django.conf import settings
+from django.core.validators import RegexValidator
+from django.db import models
+from django.template import engines
+from django.template.exceptions import TemplateDoesNotExist
+from django.template.loader import get_template
+from django.utils import six
+from django.utils.translation import ugettext_lazy as _
+
+from oscar.apps.communication.managers import CommunicationTypeManager
+from oscar.core.compat import AUTH_USER_MODEL
+from oscar.models.fields import AutoSlugField
+
+
+class AbstractEmail(models.Model):
+    """
+    This is a record of an email sent to a customer.
+    """
+    user = models.ForeignKey(
+        AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        related_name='emails',
+        verbose_name=_("User"),
+        null=True)
+    email = models.EmailField(_('Email Address'), null=True, blank=True)
+    subject = models.TextField(_('Subject'), max_length=255)
+    body_text = models.TextField(_("Body Text"))
+    body_html = models.TextField(_("Body HTML"), blank=True)
+    date_sent = models.DateTimeField(_("Date Sent"), auto_now_add=True)
+
+    class Meta:
+        abstract = True
+        app_label = 'communication'
+        db_table = 'communication_email'
+        verbose_name = _('Email')
+        verbose_name_plural = _('Emails')
+
+    def __str__(self):
+        if self.user:
+            return _(u"Email to %(user)s with subject '%(subject)s'") % {
+                'user': self.user.get_username(), 'subject': self.subject}
+        else:
+            return _(u"Anonymous email to %(email)s with subject '%(subject)s'") % {
+                'email': self.email, 'subject': self.subject}
+
+
+class AbstractCommunicationEventType(models.Model):
+    """
+    A 'type' of communication.  Like an order confirmation email.
+    """
+
+    #: Code used for looking up this event programmatically.
+    # e.g. PASSWORD_RESET. AutoSlugField uppercases the code for us because
+    # it's a useful convention that's been enforced in previous Oscar versions
+    code = AutoSlugField(
+        _('Code'), max_length=128, unique=True, populate_from='name',
+        separator=six.u("_"), uppercase=True, editable=True,
+        validators=[
+            RegexValidator(
+                regex=r'^[a-zA-Z_][0-9a-zA-Z_]*$',
```

Comment:
In the code that fetches email templates, we normalise the code with `code = self.code.lower()`. This means that if someone has defined two identical codes with different case, things will break. This seems like a bug even with the current code.

Since this is a breaking change I would suggest we enforce uppercase codes (I'd personally prefer lower, but historically Oscar has generated uppercase).